<a href="https://colab.research.google.com/github/VictorPortolani/am-turma5/blob/main/Atividade_aula0901.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pipeline de Análise de Dados [Census]


## Importação de Bibliotecas

* Pandas
* Numpy
* MatPlot
* Seaborn

In [86]:
!pip install pandas
!pip install numpy
!pip install matplotlib
!pip install seaborn

In [87]:
import pandas as pd # organizar arrays
import numpy as np #funções matemáticas complexos
import matplotlib.pyplot as plt #plotar funções matemáticas complexas
import seaborn as sns # vizualizar gráficos
import plotly.express as px #expor a biblioteca

In [88]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [89]:
base_census = pd.read_csv('/content/drive/MyDrive/am-t5/census.csv', skipinitialspace=True)


## Carregamento dos Dados da Base

In [90]:
base_census

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32557,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32558,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
32559,22,Private,201490,HS-grad,9,Never-married,Adm-clerical,Own-child,White,Male,0,0,20,United-States,<=50K


In [91]:
base_census.describe()
#income: renda
#loan: divida
#defalut: sim ou não para dívida
#std: desvio padrão
#mean: média

,age,final-weight,education-num,capital-gain,capital-loos,hour-per-week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


## Características da Base
- População  amostral: 32561
- Minorias Importantes: idade = 28 e 90, ed. = 9 e 16, ht trabalhadas = 40 e 99
- Media amostral: idade = 38.58, ed. = 10, ganho = 1077.65, perca = 87.30, ht = 40.44
- Desvio padrão [limites]: idade = 24.94, ed. = 7.51, ganho = -6757.64,
perda = -315.66, ht = 28.9   

In [92]:
grafic_class_wa = px.treemap(base_census, path=['workclass', 'age'])
grafic_class_wa.show()

In [93]:
grafic_class_ora = px.treemap(base_census, path=['occupation', 'relationship', 'age'])
grafic_class_ora.show()

In [94]:
grafic_finetune_oeg = px.parallel_categories(base_census, dimensions=['education', 'race', 'income'])
grafic_finetune_oeg.show()

In [95]:
grafic_finetune_owg = px.parallel_categories(base_census, dimensions=['workclass', 'race', 'occupation'])
grafic_finetune_owg.show()

##Análise de Anomalias de Dados [Census]
**Enunciado: Auditoria e Higienização de Dados Censitários**

Vamos analizar e estruturar a ingestão de um novo pipeline de dados sociodemográficos focado no desenvolvimento de modelos de inteligência artificial. O arquivo **census.csv** foi extraído de um sistema legado, mas o processo automatizado falhou antes mesmo da etapa de modelagem estatística. Como futuros cientistas da computação e engenheiros de dados, a missão de vocês é realizar uma auditoria rigorosa neste arquivo para garantir a integridade da informação.

Dicas de Investigação e Solução

* **Identificação de Linhas Infiltradas**: Ao forçarem o carregamento inicial (pulando as linhas defeituosas do topo), observem o comportamento dos tipos de dados informados pelo `.dtypes`. Se colunas que deveriam ser numéricas (`age` ou `hour-per-week`) foram lidas como `object`, existem *strings* perdidas no meio da massa de dados. O cabeçalho e fragmentos de texto corrompido estão repetidos ao longo do arquivo.
* **Inconsistência Categórica e Caracteres de Escape:** Analisem a distribuição de classes na coluna `income` através do `.value_counts()`. A mesma classe de renda foi acidentalmente dividida em múltiplas categorias distintas devido a falhas de exportação com caracteres de escape textuais (como `\<=50K` versus `<=50K`).
* **Dados Faltantes Não Convencionais:** Avaliem como a ausência de informação foi mascarada pelo sistema de origem. Em bases censitárias clássicas, campos em branco (como `workclass` e `native-country`) frequentemente não aparecem como `NaN` estruturados, mas são substituídos por marcadores categóricos arbitrários, como o caractere `?`.
* **Tratamento de Espaçamento:** A separação por vírgulas pode conter espaços em branco residuais no início das palavras, prejudicando filtros e agrupamentos futuros. O parâmetro `skipinitialspace=True` na leitura do Pandas é um atalho eficiente para padronizar as bordas do texto.

In [96]:
base_census.dtypes
# não há objetos em valores que deveriam ser numéricos

,0
age,int64
workclass,object
final-weight,int64
education,object
education-num,int64
marital-status,object
occupation,object
relationship,object
race,object
sex,object


In [97]:
base_census.loc[base_census.isnull().any(axis=1)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


In [98]:
colunas_texto = base_census.select_dtypes(include=['object'])

# Aplica o .strip() para limpar espaços e depois compara com "?"
(colunas_texto.apply(lambda x: x.str.strip()) == "?").sum()

,0
workclass,1836
education,0
marital-status,0
occupation,1843
relationship,0
race,0
sex,0
native-country,583
income,0


In [99]:
# Filtra as linhas onde o valor " ?" aparece em pelo menos uma das colunas
base_census.loc[(base_census == ' ?').any(axis=1)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


In [100]:
base_census.loc[base_census['workclass'].str.contains(' ?', regex=False, na=False)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


In [101]:
base_census.loc[((base_census == ' ?').any(axis=1)) & (base_census['age'] < 24)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


Normalização de Estudantes para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Dependent` e `Student`.

In [102]:
def  normalizarEstudantes(df):
  condition_age_24 = df['age'] < 24
  df.loc[condition_age_24 & (df['workclass'] == ' ?'), 'workclass'] = 'Dependent'
  df.loc[condition_age_24 & (df['occupation'] == ' ?'), 'occupation'] = 'Student'
  return df
base_census = normalizarEstudantes(base_census)

In [103]:
base_census.loc[base_census['age'] < 18]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
106,17,?,304873,10th,6,Never-married,?,Own-child,White,Female,34095,0,32,United-States,<=50K
209,17,Private,65368,11th,7,Never-married,Sales,Own-child,White,Female,0,0,12,United-States,<=50K
262,17,Private,245918,11th,7,Never-married,Other-service,Own-child,White,Male,0,0,12,United-States,<=50K
271,17,Private,191260,9th,5,Never-married,Other-service,Own-child,White,Male,1055,0,24,United-States,<=50K
335,17,Private,270942,5th-6th,3,Never-married,Other-service,Other-relative,White,Male,0,0,48,Mexico,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
31772,17,?,256173,10th,6,Never-married,?,Own-child,White,Female,0,0,15,United-States,<=50K
31864,17,Private,214787,12th,8,Never-married,Adm-clerical,Own-child,White,Female,0,0,25,United-States,<=50K
31959,17,Private,24090,HS-grad,9,Never-married,Exec-managerial,Own-child,White,Female,0,0,35,United-States,<=50K
32282,17,Private,311907,11th,7,Never-married,Other-service,Own-child,White,Male,0,0,25,United-States,<=50K


In [104]:
base_census.loc[((base_census == ' ?').any(axis=1)) & (base_census['age'] >= 62)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


- Normalização de Idosos para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Retired` e `None`.
- Normalização de Desempregados para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Unemployed` e `None`.

In [105]:
def normalizarCenso(df):
    # 1. Tratar Aposentados (62 anos ou mais)
    condicao_aposentado = df['age'] >= 62
    df.loc[condicao_aposentado & (df['workclass'] == ' ?'), 'workclass'] = ' Retired'
    df.loc[condicao_aposentado & (df['occupation'] == ' ?'), 'occupation'] = ' None'

    # 2. Tratar Desempregados (Entre 24 e 61 anos)
    condicao_desempregado = (df['age'] >= 24) & (df['age'] < 62)
    df.loc[condicao_desempregado & (df['workclass'] == ' ?'), 'workclass'] = ' Unemployed'
    df.loc[condicao_desempregado & (df['occupation'] == ' ?'), 'occupation'] = ' None'

    return df

base_census = normalizarCenso(base_census)

In [106]:
base_census.loc[base_census['age'] >= 62]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
74,79,Private,124744,Some-college,10,Married-civ-spouse,Prof-specialty,Other-relative,White,Male,0,0,20,United-States,<=50K
77,67,?,212759,10th,6,Married-civ-spouse,?,Husband,White,Male,0,0,2,United-States,<=50K
100,76,Private,124191,Masters,14,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,40,United-States,>50K
144,70,Private,105376,Some-college,10,Never-married,Tech-support,Other-relative,White,Male,0,0,40,United-States,<=50K
148,64,Private,181232,11th,7,Married-civ-spouse,Craft-repair,Husband,White,Male,0,2179,40,United-States,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32515,66,Federal-gov,47358,10th,6,Married-civ-spouse,Craft-repair,Husband,White,Male,3471,0,40,United-States,<=50K
32525,81,?,120478,Assoc-voc,11,Divorced,?,Unmarried,White,Female,0,0,1,?,<=50K
32539,71,?,287372,Doctorate,16,Married-civ-spouse,?,Husband,White,Male,0,0,10,United-States,>50K
32542,72,?,129912,HS-grad,9,Married-civ-spouse,?,Husband,White,Male,0,0,25,United-States,<=50K


In [107]:
base_census.loc[((base_census == ' Unemployed').any(axis=1)) & (base_census['age'] < 62) & (base_census['age'] >= 24)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


In [108]:
base_census.loc[((base_census == ' ?').any(axis=1))]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


### Tratamento de `native-country`

In [109]:
print(base_census['native-country'].value_counts())

native-country
United-States                 29170
Mexico                          643
?                               583
Philippines                     198
Germany                         137
Canada                          121
Puerto-Rico                     114
El-Salvador                     106
India                           100
Cuba                             95
England                          90
Jamaica                          81
South                            80
China                            75
Italy                            73
Dominican-Republic               70
Vietnam                          67
Guatemala                        64
Japan                            62
Poland                           60
Columbia                         59
Taiwan                           51
Haiti                            44
Iran                             43
Portugal                         37
Nicaragua                        34
Peru                             31
France       

In [110]:
paises_por_etnia = base_census[base_census['native-country'] != ' ?'].groupby('race')['native-country'].agg(lambda x: x.mode()[0])

print(paises_por_etnia)

race
Amer-Indian-Eskimo    United-States
Asian-Pac-Islander    United-States
Black                 United-States
Other                 United-States
White                 United-States
Name: native-country, dtype: object


In [111]:
# Verifica a distribuição de renda das pessoas SEM país declarado (os 583 casos)
# O normalize=True transforma os valores em porcentagem
renda_sem_pais = base_census[base_census['native-country'] == ' ?']['income'].value_counts(normalize=True) * 100

# Verifica a distribuição de renda das pessoas COM país declarado (o resto do dataset)
renda_com_pais = base_census[base_census['native-country'] != ' ?']['income'].value_counts(normalize=True) * 100

print("Distribuição de renda - SEM país declarado:")
print(renda_sem_pais.round(2))
print("\nDistribuição de renda - COM país declarado:")
print(renda_com_pais.round(2))

Distribuição de renda - SEM país declarado:
Series([], Name: proportion, dtype: float64)

Distribuição de renda - COM país declarado:
income
<=50K    75.92
>50K     24.08
Name: proportion, dtype: float64


In [112]:
# Verifica a distribuição de renda no dataset inteiro
renda_total = base_census['income'].value_counts(normalize=True) * 100

print("Distribuição de renda - Dataset Inteiro:")
print(renda_total.round(2))

Distribuição de renda - Dataset Inteiro:
income
<=50K    75.92
>50K     24.08
Name: proportion, dtype: float64


In [113]:
print(base_census['income'].value_counts())
#Verificação se a renda não tem anomalia (Status: OK)

income
<=50K    24720
>50K      7841
Name: count, dtype: int64


In [114]:
# Filtra apenas o grupo com país ' ?'
sem_pais = base_census[base_census['native-country'] == ' ?']

print("--- Raça dos não declarados ---")
print(sem_pais['race'].value_counts())

print("\n--- Classe de trabalho dos não declarados ---")
print(sem_pais['workclass'].value_counts())

--- Raça dos não declarados ---
Series([], Name: count, dtype: int64)

--- Classe de trabalho dos não declarados ---
Series([], Name: count, dtype: int64)


As anomalias de países não podem ser excluidas pois os dados são os seguintes, asiátios no total era um número baixo, mas nesses em que o `?` aparece, asiáticos sobem para `14.24%` e negros para `16.47%`, ou seja esses números representam minorias altas, ocasionalmente interferindo no aprendizado da máquina, solução mais propripcía substituir os países em que aparece `?` para `Unkdowm`.

In [115]:
base_census['native-country'] = base_census['native-country'].replace(' ?', ' Unknown')

In [116]:
base_census.loc[((base_census == ' ?').any(axis=1))]
#Anomalias corrigidas

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


### Resutado final
Normalização das Anomalias:
- Para dados com `age < 24` os valores de `workclass` e `ocupation` que estavam com `?` foi assumido `Dependent` e `Student` respctivamente.
- Para dados com `24 <= age < 62` os valores de `workclass` e `ocupation` que estavam com `?` foi assumido `Unemployed` e `None` respctivamente.
- Para dados com`age >= 62` os valores de `workclass` e `ocupation` que estavam com `?` foi assumido `Retired` e `None` respctivamente.
- A coluna `income` for feita a verificação e não valores declarados para múltiplas categorias.
- Para os `native-country` foi feito o seguinte, análise perante, a classe de trabalho e a etinia. No caso, por conta das etinias que representavam menores percentuais na tabela total, elas obtiveram aumento em relação aos dados com `?`, exemplo: `black` apresenta percentual total de `9.375%`, porém separando com o `native-country` representa `16.47%` desses dados, concluindo, ele se encontra no grupo das minorias, mas seu valor é alto, visto que representa maioria dentro do grupo.